pydantic/type_adapter.py

```diff
@@ -233,100 +159,203 @@ def __init__(
+            v = TypeAdapter(OuterDict)
+            v({'x': 1})  # should fail but doesn't
+            ```
+
+            If `OuterDict` were a `BaseModel`, this would work because it would resolve
+            the forward reference within the `a.py` namespace.
+            But `TypeAdapter(OuterDict)` can't determine what module `OuterDict` came from.
+
+            In other words, the assumption that _all_ forward references exist in the
+            module we are being called from is not technically always true.
+            Although most of the time it is and it works fine for recursive models and such,
+            `BaseModel`'s behavior isn't perfect either and _can_ break in similar ways,
+            so there is no right or wrong between the two.
+
+            But at the very least this behavior is _subtly_ different from `BaseModel`'s.
+        """
+        if not force and self._defer_build:
+            _mock_val_ser.set_type_adapter_mocks(self, str(self._type))
+            self.pydantic_complete = False
+            return False
+
         try:
-            self._core_schema = _getattr_no_parents(self._type, '__pydantic_core_schema__')
-            self._validator = _getattr_no_parents(self._type, '__pydantic_validator__')
-            self._serializer = _getattr_no_parents(self._type, '__pydantic_serializer__')
+            self.core_schema = _getattr_no_parents(self._type, '__pydantic_core_schema__')
+            self.validator = _getattr_no_parents(self._type, '__pydantic_validator__')
+            self.serializer = _getattr_no_parents(self._type, '__pydantic_serializer__')
+
+            # TODO: we don't go through the rebuild logic here directly because we don't want
+            # to repeat all of the namespace fetching logic that we've already done
+            # so we simply skip to the block below that does the actual schema generation
+            if (
+                isinstance(self.core_schema, _mock_val_ser.MockCoreSchema)
+                or isinstance(self.validator, _mock_val_ser.MockValSer)
+                or isinstance(self.serializer, _mock_val_ser.MockValSer)
+            ):
+                raise AttributeError()
         except AttributeError:
             config_wrapper = _config.ConfigWrapper(self._config)
+
+            schema_generator = _generate_schema.GenerateSchema(config_wrapper, ns_resolver=ns_resolver)
+
+            try:
+                core_schema = schema_generator.generate_schema(self._type)
+            except PydanticUndefinedAnnotation:
+                if raise_errors:
+                    raise
+                _mock_val_ser.set_type_adapter_mocks(self, str(self._type))
+                return False
+
+            try:
+                self.core_schema = schema_generator.clean_schema(core_schema)
+            except schema_generator.CollectedInvalid:
+                _mock_val_ser.set_type_adapter_mocks(self, str(self._type))
+                return False
```

Comment:
I have seen the same try-catch-if-raise_error-mocking logic repeated in many places. Wondering could it be extracted to some common handling... Maybe later...

Reply:
Yeah, let's consolidate later

Reply:
Maybe with https://github.com/pydantic/pydantic/commit/341a778521987f9a9bc91cf58c930e3c2fdb55e4, I can create an issue

Reply:
https://github.com/pydantic/pydantic/issues/10831